### PR TITLE
[SPARK-16344][SQL] Decoding Parquet array of struct with a single field named "element"

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

Due to backward-compatibility reasons, the following Parquet schema is ambiguous:

```
optional group f (LIST) {
  repeated group list {
    optional group element {
      optional int32 element;
    }
  }
}
```

According to the parquet-format spec, when interpreted as a standard 3-level layout, this type is equivalent to the following SQL type:

```
ARRAY<STRUCT<element: INT>>
```

However, when interpreted as a legacy 2-level layout, it's equivalent to

```
ARRAY<STRUCT<element: STRUCT<element: INT>>>
```

Historically, to disambiguate these cases, we employed two methods:
- `ParquetSchemaConverter.isElementType()`
  
  Used to disambiguate the above cases while converting Parquet types to Spark types.
- `ParquetRowConverter.isElementType()`
  
  Used to disambiguate the above cases while instantiating row converters that convert Parquet records to Spark rows.

Unfortunately, these two methods make a different decision about the above problematic Parquet type and caused SPARK-16344.

`ParquetRowConverter.isElementType()` is necessary for Spark 1.4 and earlier versions because Parquet requested schemata are directly converted from Spark schemata in these versions. The converted Parquet schemata may be incompatible with actual schemata of the underlying physical files when the files are written by a system/library that uses a schema conversion scheme that is different from Spark when writing Parquet LIST and MAP fields.

In Spark 1.5, Parquet requested schemata are always properly tailored from schemata of physical files to be read. Thus `ParquetRowConverter.isElementType()` is no longer necessary. This PR replaces this method with a simple yet accurate scheme: whenever an ambiguous Parquet type is hit, convert the type in question back to a Spark type using `ParquetSchemaConverter` and check whether it matches the corresponding Spark type.
## How was this patch tested?

New test cases added in `ParquetHiveCompatibilitySuite` and `ParquetQuerySuite`.

### CODE REVIEW

**Comment by liancheng** [comment]
cc @yhuai

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala [hunk 1]
```diff
@@ -482,13 +482,106 @@ private[parquet] class ParquetRowConverter(
      */
     // scalastyle:on
     private def isElementType(
-        parquetRepeatedType: Type, catalystElementType: DataType, parentName: String): Boolean = {
+        parquetRepeatedType: Type, catalystElementType: DataType, parent: GroupType): Boolean = {
+
+      def isStandardListLayout(t: GroupType): Boolean =
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: can we inline this in the pattern match

**Comment by rxin** [comment]
Let's not merge this in 2.0, unless we can get more than one person to understand the patch...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala (hunk 1 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: Done.

**Comment by liancheng** [comment]
The last build failure seems to be caused by flaky tests.

**Comment by liancheng** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala [hunk 2]
```diff
@@ -482,13 +482,104 @@ private[parquet] class ParquetRowConverter(
...
+          //    // List<Tuple<String, Integer>> (nullable list, non-null elements)
+          //    optional group my_list (LIST) {           <-- parent
+          //      repeated group element {                <-- repeatedType
+          //        required binary str (UTF8);
+          //        required int32 num;
+          //      };
+          //    }
+          true
+
+        case (t: GroupType, _) if t.getFieldCount == 1 && t.getName == "array" =>
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: The below comment doesn't match this condition (`t.getName == "array"`)?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: Nice catch!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala [hunk 3]
```diff
@@ -482,13 +482,105 @@ private[parquet] class ParquetRowConverter(
...
+          // e.g.:
+          //
+          //    // List<OneTuple<String>> (nullable list, non-null elements)
+          //    optional group my_list (LIST) {           <-- parent
+          //      repeated group list {                   <-- repeatedType
+          //        required binary str (UTF8);
+          //      };
+          //    }
+          //
+          // NOTE: This kind of schema is ambiguous. According to parquet-format spec, this schema
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: This few sentences seems can be improved to be more clear. Like: According to parquet-format spec, this schema can also be interpreted as `List<String>` if following standard 3-level list types as shown in above case branch. In order to avoid such ambiguous interpretation, we prefer standard 3-level list types and place this case after the above one. ...

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: Yea, the current comment is kinda contrived... Let me try again.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala [hunk 4]
```diff
@@ -482,13 +482,105 @@ private[parquet] class ParquetRowConverter(
...
-        case (t: GroupType, _) if t.getFieldCount > 1 => true
-        case (t: GroupType, _) if t.getFieldCount == 1 && t.getName == "array" => true
-        case (t: GroupType, _) if t.getFieldCount == 1 && t.getName == parentName + "_tuple" => true
-        case (t: GroupType, StructType(Array(f))) if f.name == t.getFieldName(0) => true
+        case (t: PrimitiveType, _) =>
+          // For legacy 2-level list types with primitive element type, e.g.:
+          //
+          //    // List<Integer> (nullable list, non-null elements)
+          //    optional group my_list (LIST) {           <-- parent
+          //      repeated int32 element;                 <-- repeatedType
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: `<-- repeatedType` => `<-- parquetRepeatedType`?

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: Seems it is better to make the name match the variable name.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala [hunk 5]
```diff
@@ -482,13 +482,105 @@ private[parquet] class ParquetRowConverter(
...
+
+        case (t: GroupType, _) if t.getFieldCount > 1 =>
+          // For legacy 2-level list types whose element type is a group type with 2 or more fields,
+          // e.g.:
+          //
+          //    // List<Tuple<String, Integer>> (nullable list, non-null elements)
+          //    optional group my_list (LIST) {           <-- parent
+          //      repeated group element {                <-- repeatedType
+          //        required binary str (UTF8);
+          //        required int32 num;
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: Also mention that `t.getFieldCount` is 2 (including `str` and `num`) for this case?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala [hunk 6]
```diff
@@ -482,13 +482,105 @@ private[parquet] class ParquetRowConverter(
...
+          //      repeated group list {                   <-- repeatedType
+          //        required binary element (UTF8);
+          //      }
+          //    }
+          //
+          // This case branch must appear before the next one. See comments of the next case branch
+          // for details.
+          false
+
+        case (t: GroupType, StructType(fields)) =>
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: So, when we reach here, we have `t.getFieldCount==1`? Is it possible that `t.getFieldCount == 0`?

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: Should we have a case to explicitly handle `t.getFieldCount == 0` (to make the code easier to follow)?

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: Probably it is good to explain why we are explicitly matching `GroupType` with `StructType` (because we are trying to determine if `GroupType` represents a Struct or just a middle layer?).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala [hunk 7]
```diff
@@ -482,13 +482,105 @@ private[parquet] class ParquetRowConverter(
...
+          // For legacy 2-level list types whose element type is a group type with a single field,
+          // e.g.:
+          //
+          //    // List<OneTuple<String>> (nullable list, non-null elements)
+          //    optional group my_list (LIST) {           <-- parent
+          //      repeated group list {                   <-- repeatedType
+          //        required binary str (UTF8);
+          //      };
+          //    }
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: Also provide the Spark's corresponding StructType in this example?

**Comment by liancheng** [comment]
@yhuai I found more test cases that may fail even with changes made in this PR. A proper fix is delivered in the last commit. Details about the new test case and the new fix can be found in comment area of SPARK-16344.

**Comment by liancheng** [comment]
retest this please

**Comment by rdblue** [comment]
+1

Nice work, I think we may be able to do the same thing in parquet-avro as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala [hunk 8]
```diff
@@ -260,7 +260,7 @@ private[parquet] class ParquetSchemaConverter(
     {
       // For legacy 2-level list types with primitive element type, e.g.:
       //
-      //    // List<Integer> (nullable list, non-null elements)
+      //    // ARRAY<INT> (nullable list, non-null elements)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala]: I'm using concrete SQL types here for better readability.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRecordMaterializer.scala [hunk 9]
```diff
@@ -30,10 +30,11 @@ import org.apache.spark.sql.types.StructType
  * @param catalystSchema Catalyst schema of the rows to be constructed
  */
 private[parquet] class ParquetRecordMaterializer(
-    parquetSchema: MessageType, catalystSchema: StructType)
+    parquetSchema: MessageType, catalystSchema: StructType, schemaConverter: ParquetSchemaConverter)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRecordMaterializer.scala]: Add `schemaConverter` to the scaladoc?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala [hunk 10]
```diff
@@ -442,13 +445,23 @@ private[parquet] class ParquetRowConverter(
...
+        // If the repeated field corresponds to the element type, creates a new converter using the
+        // type of the repeated field.
         newConverter(repeatedType, elementType, new ParentContainerUpdater {
           override def set(value: Any): Unit = currentArray += value
         })
       } else {
+        // If the repeated field corresponds to the syntactic group in the standard 3-level Parquet
+        // LIST layout, creates a new converter using the only child field of the repeated field.
+        assert(!repeatedType.isPrimitive && repeatedType.asGroupType().getFieldCount == 1)
         new ElementConverter(repeatedType.asGroupType().getType(0), elementType)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala]: Can we add examples at here?

**Comment by yhuai** [comment]
Let's also update the description.

**Comment by liancheng** [comment]
@yhuai All comments addressed. PR description is also updated. Thanks for the review!

**Comment by yhuai** [comment]
Thank you! I am going to merge this to master.

**Comment by deepakmk9** [comment]
IS there any way to make this work in Spark 1.6.2 , I ma facing the same issue
